### PR TITLE
Revised and tested GraphTests in experimental

### PR DESCRIPTION
**Pull request by d-michail**
A very small utility method that checks whether a graph is simple (has no loops and multiple edges). It checks whether the actual graph is simple, not whether its type supports non-simple graphs.

This is something useful since many algorithms do not work out of the box with graphs that are not simple.

@jsichi, @jkinable I am not sure whether this or similar methods should be in the Graphs class. There is a class called GraphTests but it is in the experimental package (and also needs, some refactoring). Maybe add a GraphTests in the jgrapht package?

Moreover, if you see the implementation, it is directly related to issue #298.

### CODE REVIEW

**Comment by jkinable** [comment]
yes, let's put this in GraphTests

File: jgrapht-core/src/main/java/org/jgrapht/GraphTests.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+                }
+                if (!neighbors.add(u)) {
+                    return false;
+                }
+            }
+        }
+
+        return true;
+    }
+}
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/GraphTests.java]: What about the other tests that are present in the current experimental GraphTests? Do we want to keep those, or did you leave them out on purpose? If you left them out on purpose, we may want to rethink having a dedicated GraphTests class, unless you envision more tests to be added somewhere in the future.

File: jgrapht-core/src/main/java/org/jgrapht/GraphTests.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+            if (!abg.isAllowingLoops() && !abg.isAllowingMultipleEdges()) {
+                return true;
+            }
+        }
+        // no luck, we have to check
+        boolean isDirected = graph instanceof DirectedGraph<?, ?>;
+        for (V v : graph.vertexSet()) {
+            Set<V> neighbors = new HashSet<>();
+            Set<E> edgesOf;
+            if (isDirected) {
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/GraphTests.java]: another candidate for issue #298 :)

File: jgrapht-core/src/main/java/org/jgrapht/GraphTests.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+                edgesOf = ((DirectedGraph<V, E>) graph).outgoingEdgesOf(v);
+            } else {
+                edgesOf = graph.edgesOf(v);
+            }
+            for (E e : edgesOf) {
+                V u = Graphs.getOppositeVertex(graph, e, v);
+                if (u.equals(v)) {
+                    return false;
+                }
+                if (!neighbors.add(u)) {
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/GraphTests.java]: could you combine these 2 if statements?

File: jgrapht-core/src/main/java/org/jgrapht/GraphTests.java (hunk 1 above)
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/GraphTests.java]: I did not move them on purpose as the need a bit of refactoring and testing. 

However, I think we should add at least the following: 
 - isBipartite
 - isConnected(undirected)
 - isStronglyConnected(directed)

**Comment by d-michail** [comment]
The only method still in the experimental package is the `isBipartite()`. I hope I find some time to do this also.

**Comment by d-michail** [comment]
Checked all the tests and moved them out of experimental.

File: jgrapht-core/src/main/java/org/jgrapht/GraphTests.java [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+                allEdges = Math.multiplyExact(n, (n - 1) / 2);
+            }
+        } else {
+            throw new IllegalArgumentException("Graph must be directed or undirected");
+        }
+        return graph.edgeSet().size() == allEdges && isSimple(graph);
+    }
+
+    /**
+     * Test whether an undirected graph is connected.
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/GraphTests.java]: Could you mention that this method delegates to the ConnectivityInspector?

File: jgrapht-core/src/main/java/org/jgrapht/GraphTests.java [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+
+    /**
+     * Test whether a directed graph is weakly connected.
+     * 
+     * @param graph the input graph
+     * @param <V> the graph vertex type
+     * @param <E> the graph edge type
+     * @return true if the graph is weakly connected, false otherwise
+     */
+    public static <V, E> boolean isWeaklyConnected(DirectedGraph<V, E> graph)
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/GraphTests.java]: Same here

**Comment by jkinable** [comment]
Great work, nice usage of some java 8 functionality too, very clean. Few minor comments.

**Comment by d-michail** [comment]
Thanks for the comments! I incorporated them in the current version.

File: jgrapht-core/src/test/java/org/jgrapht/GraphTestsTest.java [hunk 6]
```diff
@@ -0,0 +1,497 @@
...
+        g.addEdge(2, 2);
+        g.addEdge(3, 3);
+        g.addEdge(4, 4);
+        g.addEdge(5, 5);
+        Assert.assertFalse(GraphTests.isEulerian(g));
+        g.addEdge(1, 5);
+        Assert.assertTrue(GraphTests.isEulerian(g));
+    }
+
+    class IntegerVertexFactory
```
Comment (d-michail) [inline jgrapht-core/src/test/java/org/jgrapht/GraphTestsTest.java]: I added a default implementation called `IntegerVertexFactory` but kept it in the test classpath.

**Comment by jkinable** [comment]
JGraphT on a diet: 17 IntegerVertexFactories lost in 1 pull request! Great work :)

**Pull request merged**